design/01-proc-macro.md

```diff
@@ -0,0 +1,290 @@
+    // no-op
+    ProcMacroResult::Leave
+}
+```
+
+## Reference-level explanation
+
+### Procedural macros will be compiled into shared libraries
+
+- Procedural macros will be implemented in Rust.
+- Procedural macros will be compiled into shared libraries, that Scarb can load at the runtime through dynamic loading mechanism of the host operating system.
+- Procedural macros will be compiled to `.dylib` files on OSX, `.so` files on Linux and `.dll` on Windows.
+- Scarb will use [`libloading` crate](https://docs.rs/libloading/latest/libloading/), which exposes a cross-platform interface for interactions with platform dynamic library loading primitives, to load shared libraries into memory.
+
+### Procedural macros will be special Scarb packages containing Rust code
+
+- A procedural macro will be a special Scarb package, which will use a new native target type: `cairo-plugin`.
+- The procedural macro package will contain Rust source code, which will be shipped to users on Scarb project resolution.
+- The procedural macro source code will be compiled on Scarb users system only.
+- Enabling this target will mean that this package does not contain any Cairo code.
+- This target will be *exclusive*:
+    - It will block defining other targets for the package, not even `lib`.
+    - It will also not be possible to declare dependencies, or specify Cairo compiler settings, it won't make sense for these packages.
+- Scarb will automatically generate a `Cargo.toml` manifest file, based on contents of `Scarb.toml` manifest. The Cargo manifest will be used for the procedural macro compilation.
+    - The `lib.crate-type` will be set `dylib`
+- The Cargo manifest generation, followed by the Cargo resolution of procedural macro packages, will be handled during the Scarb workspace resolution.
+- The compiled procedural macros will be cached on the user filesystem in Scarb cache directory, to ensure no recompilation on each Scarb project build is required.
+
+**Example:**
+
+Scarb manifest file:
+
+```toml
+[package]
+name = "some-macro"
+version = "0.1.0"
+
+[cairo-plugin]
+```
+
+Automatically generated Cargo manifest file:
+
+```toml
+# Code generated by scarb fetch; DO NOT EDIT.
+#
+# The Cargo.toml file for Cairo procedural macros crates is entirely managed by Scarb,
+# the Cairo package manager. Scarb regenerates this file whenever it processes Scarb.toml.
+# The simplest way to trigger this, is to run the `scarb fetch` command.
+#
+# If you want to provide custom data in this file, include it in Scarb.toml under
+# [tool.cargo.pkg] section. For example, to add Rust dependencies, type the following:
+#
+#     [tool.cargo.pkg.dependencies]
+#     serde = "1"
+#     serde_json = "1"
+
+[package]
+name = "ignore-plugin"
+version = "0.1.0"
+
+[lib]
+crate-type = "dylib"
+
+[dependencies]
+cairo-lang-macro = "1.0.0-alpha.7"
+```
+
+### Procedural macro packages can be used as regular dependencies
```

Comment:
These two sections are duplicated. Do you have a reasoning behind doing this? 🤔

Reply:
Just in case someone skipped guide-level proposal and went straight to the reference-level one. I will reword the former to avoid straight duplication.

Reply:
You can just make references instead of copying paragraphs

Reply:
Replaced with a reference.